### PR TITLE
CUDA 9 support

### PR DESCRIPTION
**Pull request by slayton58**
Adds support for the CUDA 9 toolkit.

Includes new fp16 data type fixes, and changes to warp-synchronous programming. Also updates CUB third-party repo for CUDA 9 support.

### CODE REVIEW

File: cmake/Cuda.cmake [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # Known NVIDIA GPU achitectures Caffe2 can be compiled for.
 # This list will be used for CUDA_ARCH_NAME = All option
-set(Caffe2_known_gpu_archs "20 21(20) 30 35 50 52 60 61")
+set(Caffe2_known_gpu_archs "20 21(20) 30 35 50 52 60 61 70")
```
Comment (pietern) [inline cmake/Cuda.cmake:3]: Can you add `Caffe2_known_gpu_archs8` that doesn't include 70, like the variant for CUDA 7 below?

**Review by pietern — COMMENTED** [review]
Nice! Thanks, @slayton58. This looks very isolated indeed.

File: cmake/Cuda.cmake (hunk 1 above)
Comment (slayton58) [reply cmake/Cuda.cmake:3]: Updated

**Comment by yapus** [comment]
I've just tried to build this branch in conda 4.3.22 under Mac OS X 10.12.6, Cuda 9.0.103, CudNN-9.0 osx-x64-v7 and I get the following errors:

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory(2577): error: calling a __host__ function("std::__1::__compressed_pair< ::cub::CachingDeviceAllocator *,  ::std::__1::default_delete< ::cub::CachingDeviceAllocator> > ::__compressed_pair") from a __device__ function("std::__1::unique_ptr< ::cub::CachingDeviceAllocator,  ::std::__1::default_delete< ::cub::CachingDeviceAllocator> > ::unique_ptr") is not allowed

/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory(2577): error: identifier "std::__1::__compressed_pair< ::cub::CachingDeviceAllocator *,  ::std::__1::default_delete< ::cub::CachingDeviceAllocator> > ::__compressed_pair" is undefined in device code
```

**Comment by yapus** [comment]
Full build log if it helps: https://pastebin.com/sKDhkTYg

**Comment by slayton58** [comment]
Can you try updating third_party/cub to the latest upstream master?

**Comment by yapus** [comment]
I've tried with latest third_party/cub, but still can't get it working:

```
caffe2/caffe2/utils/conversions.h(166): warning: calling a __host__ function from a __host__ __device__ function is not allowed

caffe2/caffe2/core/operator_schema.h(277): warning: statement is unreachable

caffe2/caffe2/core/operator_schema.h(277): warning: statement is unreachable

caffe2/caffe2/core/operator_schema.h(277): warning: statement is unreachable

caffe2/caffe2/utils/conversions.h(166): warning: calling a __host__ function from a __host__ __device__ function is not allowed

caffe2/caffe2/core/operator_schema.h(277): warning: statement is unreachable

caffe2/caffe2/core/operator_schema.h(277): warning: statement is unreachable

caffe2/caffe2/utils/conversions.h(166): warning: calling a __host__ function from a __host__ __device__ function is not allowed

/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory(2577): error: calling a __host__ function("std::__1::__compressed_pair< ::cub::CachingDeviceAllocator *,  ::std::__1::default_delete< ::cub::CachingDeviceAllocator> > ::__compressed_pair") from a __device__ function("std::__1::unique_ptr< ::cub::CachingDeviceAllocator,  ::std::__1::default_delete< ::cub::CachingDeviceAllocator> > ::unique_ptr") is not allowed

/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory(2577): error: identifier "std::__1::__compressed_pair< ::cub::CachingDeviceAllocator *,  ::std::__1::default_delete< ::cub::CachingDeviceAllocator> > ::__compressed_pair" is undefined in device code

2 errors detected in the compilation of "/var/folders/t3/32fl0c6j2cb8d4hprb66lpth0000gn/T//tmpxft_0000ec83_00000000-6_context_gpu.cpp1.ii".
CMake Error at Caffe2_GPU_generated_context_gpu.cu.o.Release.cmake:278 (message):
  Error generating file
  caffe2/build/caffe2/CMakeFiles/Caffe2_GPU.dir/core/./Caffe2_GPU_generated_context_gpu.cu.o


make[2]: *** [caffe2/CMakeFiles/Caffe2_GPU.dir/core/Caffe2_GPU_generated_context_gpu.cu.o] Error 1
make[2]: *** Waiting for unfinished jobs....
caffe2/caffe2/core/operator_schema.h(277): warning: statement is unreachable

caffe2/caffe2/core/operator_schema.h(277): warning: statement is unreachable

caffe2/caffe2/core/operator_schema.h(277): warning: statement is unreachable

caffe2/caffe2/core/operator_schema.h(277): warning: statement is unreachable

caffe2/caffe2/core/operator_schema.h(277): warning: statement is unreachable

make[1]: *** [caffe2/CMakeFiles/Caffe2_GPU.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by slayton58** [comment]
@yapus Just to confirm, can you compile successfully with CUDA 8?

**Comment by yapus** [comment]
@slayton58 there's no CudNN library for Cuda8 on MacOS available from nVidia... That's why i didn't give it a try. Doesn't caffe2 require CudNN along with Cuda ?

**Comment by slayton58** [comment]
You could try with cudnn v6 instead of v7

**Comment by yapus** [comment]
btw when i try with caffe2@master + third_party/cub@master – it fails the other way:

```
caffe2/caffe2/perfkernels/typed_axpy_avx2.cc:19:15: error: use of undeclared identifier '_cvtsh_ss'
    *(y++) += _cvtsh_ss((*(x++)).x) * a;
              ^
caffe2/caffe2/perfkernels/typed_axpy_avx.cc:19:15: error: use of undeclared identifier '_cvtsh_ss'
    *(y++) += _cvtsh_ss((*(x++)).x) * a;
              ^
caffe2/caffe2/perfkernels/typed_axpy_avx2.cc:40:21: error: use of undeclared identifier '_cvtsh_ss'
      y[current] += _cvtsh_ss(x[current].x) * a;
                    ^
caffe2/caffe2/perfkernels/typed_axpy_avx.cc:41:21: error: use of undeclared identifier '_cvtsh_ss'
      y[current] += _cvtsh_ss(x[current].x) * a;
                    ^
2 errors generated.
2 errors generated.
```

**Comment by slayton58** [comment]
That at least shouldn't be anything to do with CUDA - it's AVX2 instructions being compiled purely via the host compiler.

**Comment by yapus** [comment]
@slayton58 Ok, i'll try using Cuda8 and CudNN6, thanks...

**Comment by yapus** [comment]
My `__clang_major__` is 8:
```
$ echo | clang -dM -E - | grep __clang_
#define __clang__ 1
#define __clang_major__ 8
#define __clang_minor__ 0
#define __clang_patchlevel__ 0
#define __clang_version__ "8.0.0 (clang-800.0.42.1)"
```

To overcome `error: use of undeclared identifier '_cvtsh_ss'` – I had to tweak it like this:
```
diff --git a/caffe2/perfkernels/cvtsh_ss_bugfix.h b/caffe2/perfkernels/cvtsh_ss_bugfix.h
index bd06681..d421f4b 100644
--- a/caffe2/perfkernels/cvtsh_ss_bugfix.h
+++ b/caffe2/perfkernels/cvtsh_ss_bugfix.h
@@ -1,6 +1,6 @@
 #pragma once
 
-#if defined(__APPLE__) && (__clang_major__ < 8)
+#if defined(__APPLE__) && (__clang_major__ < 9)
 
 #include <emmintrin.h>
 
```

I'll give a try to Cuda8 and report back here.

**Comment by yapus** [comment]
@slayton58 Thanks, It builds with Cuda-8 and CudNN-6
But i'm still curious why i couldn't make it with Cuda-9 + CudNN-7. If you have any ideas what else i could try - please let me know

**Pull request closed without merging**